cpmpy/solvers/pindakaas.py

```diff
@@ -0,0 +1,313 @@
+        return cpm_cons
+
+    def __add__(self, cpm_expr_orig):
+      """
+            Eagerly add a constraint to the underlying solver.
+
+            Any CPMpy expression given is immediately transformed (through `transform()`)
+            and then posted to the solver in this function.
+
+            This can raise 'NotImplementedError' for any constraint not supported after transformation
+
+            The variables used in expressions given to add are stored as 'user variables'. Those are the only ones
+            the user knows and cares about (and will be populated with a value after solve). All other variables
+            are auxiliary variables created by transformations.
+
+            What 'supported' means depends on the solver capabilities, and in effect on what transformations
+            are applied in `transform()`.
+
+      """
+      import pindakaas as pkl
+      if self.unsatisfiable:
+          return self
+
+      # add new user vars to the set
+      get_variables(cpm_expr_orig, collect=self.user_vars)
+
+      # transform and post the constraints
+      try: 
+          for cpm_expr in self.transform(cpm_expr_orig):
+              if cpm_expr.name == 'or':
+                  self.pkl_solver.add_clause(self.solver_vars(cpm_expr.args))
+
+              elif cpm_expr.name == '->':  # BV -> BE only thanks to only_bv_reifies
+                  a0,a1 = cpm_expr.args
+                  self._add_bool_linear(a1, conditions=[~a0])
+
+              elif isinstance(cpm_expr, Comparison):
+                  self._add_bool_linear(cpm_expr)
+
+              elif isinstance(cpm_expr, BoolVal):
+                  # base case: Boolean value
+                  if cpm_expr.args[0] is False:
+                      self.pkl_solver.add_clause([])
+
+              elif isinstance(cpm_expr, _BoolVarImpl):
+                  # base case, just var or ~var
+                  self.pkl_solver.add_clause([self.solver_var(cpm_expr)])
+
+              # a direct constraint, pass to solver
+              elif isinstance(cpm_expr, DirectConstraint):
+                  raise NotImplementedError(f"TODO")
+                  cpm_expr.callSolver(self, self.pysat_solver)
+
+              else:
+                raise NotImplementedError(f"{self.name}: Non supported constraint {cpm_expr}")
+      except pkl.Unsatisfiable:
+          self.unsatisfiable = True
+
+      return self
+
+
+    """ Unpack implied literal, clause, sum, or weighted sum """
+    def _add_bool_linear(self, cpm_expr, conditions=[]):
+        import pindakaas as pkl
+        literals = None
+        coefficients = None
+        comparator = None
+        k = None
+        if isinstance(cpm_expr, _BoolVarImpl):
+            literals = [cpm_expr]
+        elif isinstance(cpm_expr, Operator) and cpm_expr.name == 'or':
+            literals = cpm_expr.args
+        elif hasattr(cpm_expr, 'decompose'):
+            self += a0.implies(cpm_expr.decompose())
```

Comment:
a0 seems undefined, I think you just want to self += expr.decomp()

Reply:
also, this won't work well if 'conditions' is not empty?

Reply:
I believe this attribute is the responsibility of the transformations, so I've removed it.